lib/rules/prefer-destructuring.js

```diff
@@ -110,22 +167,27 @@ module.exports = {
             }
 
             if (isArrayIndexAccess(rightNode)) {
-                if (checkArrays) {
+                if ((shouldCheck("VariableDeclarator", "array") && reportNode.type === "VariableDeclarator") ||
+                    (shouldCheck("AssignmentExpression", "array") && reportNode.type === "AssignmentExpression") ||
```

Comment:
The main reason why I added the reportNode.type check is to handle cases
```
        {
            code: "foo = array[0];",
            options: [{ VariableDeclarator: { array: true }, AssignmentExpression: { array: false } }, { enforceForRenamedProperties: false }]
        },
        {
            code: "var foo = array[0];",
            options: [{ VariableDeclarator: { array: false }, AssignmentExpression: { array: true } }, { enforceForRenamedProperties: false }]
        },

```
where the check for array/object is swapped between variabledeclarator and assignmentexpression